### PR TITLE
Support completions endpoints

### PR DESCRIPTION
**Pull request by prithvikannan**
### Related Issues/PRs

 #xxx

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @prithvikannan !

File: mlflow/metrics/genai/model_utils.py [hunk 1]
```diff
@@ -115,41 +123,78 @@ def _call_openai_api(openai_uri, payload):
...
-                "metadata": {"finish_reason": c["finish_reason"]},
-            }
-            for c in resp["choices"]
-        ],
-    }
+def _call_gateway_api(gateway_uri, payload, eval_parameters):
+    from mlflow.gateway import get_route, query
+
+    route_info = get_route(gateway_uri).dict()
+    if route_info["route_type"] == "llm/v1/completions":
```
Comment (harupy) [inline mlflow/metrics/genai/model_utils.py:173]:
```suggestion
    if route_info["endpoint_type"] == "llm/v1/completions":
```

can we use `endpoint_type`?

File: mlflow/metrics/genai/model_utils.py [hunk 2]
```diff
@@ -67,7 +69,13 @@ def _call_openai_api(openai_uri, payload):
         if getattr(api_config, x) is not None
     }
 
-    payload = _convert_to_openai_request(payload)
+    payload = {
+        "prompt": payload,
+        **{{"candidate_count": "n"}.get(k, k): v for k, v in eval_parameters.items()},
```
Comment (harupy) [inline mlflow/metrics/genai/model_utils.py:74]: Do we still need this conversion?

File: mlflow/metrics/genai/model_utils.py [hunk 3]
```diff
@@ -67,7 +69,13 @@ def _call_openai_api(openai_uri, payload):
...
+    }
+    # The range of OpenAI's temperature is 0-2, but ours is 0-1, so we double it.
+    payload["temperature"] = 2 * payload["temperature"]
```
Comment (harupy) [inline mlflow/metrics/genai/model_utils.py:77]: same here: https://github.com/mlflow/mlflow/pull/10577/files#r1413275765

File: mlflow/metrics/genai/model_utils.py [hunk 4]
```diff
@@ -67,7 +69,13 @@ def _call_openai_api(openai_uri, payload):
...
+    payload["messages"] = [{"role": "user", "content": payload.pop("prompt")}]
```
[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/metrics/genai/model_utils.py:78]: not a blocker, can be addressed in a follow-up.

File: mlflow/metrics/genai/model_utils.py [hunk 5]
```diff
@@ -115,41 +123,78 @@ def _call_openai_api(openai_uri, payload):
...
+            "prompt": payload,
+            **eval_parameters,
+        }
+        response = client.predict(endpoint=deployment_uri, inputs=completions_payload)
+        try:
+            text = response["choices"][0]["text"]
+        except (KeyError, IndexError, TypeError):
+            text = None
+        return text
+    elif endpoint["task"] == "llm/v1/chat":
```
[the remark this answers is not in the document]
Comment (prithvikannan) [reply mlflow/metrics/genai/model_utils.py:150]: `MLflowDeploymentClient` return the `endpoint_task` -- can we use this?

https://github.com/mlflow/mlflow/blob/gateway-migration/mlflow/deployments/mlflow/__init__.py#L145-L163

File: mlflow/metrics/genai/model_utils.py [hunk 6]
```diff
@@ -115,41 +123,78 @@ def _call_openai_api(openai_uri, payload):
...
+            text = None
+        return text
+    elif route_info["route_type"] == "llm/v1/chat":
+        chat_payload = {
+            "messages": [{"role": "user", "content": payload}],
+            **eval_parameters,
+        }
+        response = query(gateway_uri, chat_payload)
+        try:
+            text = response["candidates"][0]["message"]["content"]
```
[the remark this answers is not in the document]
Comment (prithvikannan) [reply mlflow/metrics/genai/model_utils.py:191]: is the gateway API changing?

[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/metrics/genai/model_utils.py:191]: @prithvikannan Yeah, it's changing to match OpenAI

File: mlflow/metrics/genai/model_utils.py [hunk 7]
```diff
@@ -11,17 +11,19 @@
...
+    if eval_parameters is None:
+        eval_parameters = {}
     prefix, suffix = _parse_model_uri(model_uri)
 
     if prefix == "openai":
-        return _call_openai_api(suffix, payload)
+        return _call_openai_api(suffix, payload, eval_parameters)
     elif prefix == "gateway":
-        return _call_gateway_api(suffix, payload)
+        return _call_gateway_api(suffix, payload, eval_parameters)
```
[the remark this answers is not in the document]
Comment (prithvikannan) [reply mlflow/metrics/genai/model_utils.py:24]: Yes. Lmk if this is not needed and we can remove

File: mlflow/metrics/genai/model_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (prithvikannan) [reply mlflow/metrics/genai/model_utils.py]: Solved this by checking for `task` or `endpoint_type`

**Pull request merged**